src/pyhf/infer/intervals.py

```diff
@@ -2,13 +2,125 @@
+
+    Args:
+        data (:obj:`tensor`): The observed data.
+        model (~pyhf.pdf.Model): The statistical model adhering to the schema ``model.json``.
+        low (:obj:`float`): Lower boundary of search region
+        high (:obj:`float`): Higher boundary of search region
+        level (:obj:`float`): The threshold value to evaluate the interpolated results at.
+                              Defaults to 0.05.
+        atol (:obj:`float`): Absolute tolerance. Defaults to 1e-12. The iteration will end when the
+                             result is within absolute *or* relative tolerance of the true limit.
+        rtol (:obj:`float`): Relative tolerance. For optimal performance this argument should be set
+                             to the highest acceptable relative tolerance, though it will default
+                             to 1e-15 if not set.
+        calctype (:obj:`str`): Calculator to use for hypothesis tests. Choose 'asymptotics' (default)
+                               or 'toybased'.
+        test_stat (:obj:`str`): Test statistic to use. Choose 'qtilde' (default), 'q', or 'q0'.
+
+    Returns:
+        Tuple of Tensors:
+
+            - Tensor: The observed upper limit on the POI.
+            - Tensor: The expected upper limits on the POI.
+    """
+    if rtol is None:
+        _warn(
+            "upperlimit_auto: rtol not provided, defaulting to 1e-15. "
+            "For optimal performance rtol should be set to the highest acceptable relative tolerance."
+        )
+        rtol = 1e-15
+
+    cache = {}
+
+    def f_all(mu):
+        if mu in cache:
+            return cache[mu]
+        cache[mu] = hypotest(
+            mu,
+            data,
+            model,
+            test_stat=test_stat,
+            calctype=calctype,
+            return_expected_set=True,
+        )
+        return cache[mu]
+
+    def f(mu, limit=0):
+        # Use integers for limit so we don't need a string comparison
+        if limit == 0:
+            # Obs
+            return f_all(mu)[0] - level
+        # Exp
+        # (These are in the order -2, -1, 0, 1, 2 sigma)
+        return f_all(mu)[1][limit - 1] - level
+
+    def best_bracket(limit):
+        # return best bracket
+        ks = np.array(list(cache.keys()))
+        vals = np.array(
+            [
+                v[0] - level if limit == 0 else v[1][limit - 1] - level
+                for v in cache.values()
+            ]
+        )
+        pos = vals >= 0
+        neg = vals < 0
+        lower = ks[pos][np.argmin(vals[pos])]
+        upper = ks[neg][np.argmax(vals[neg])]
+        return (lower, upper)
+
+    tb, _ = get_backend()
+    obs = tb.astensor(_toms748(f, low, high, args=(0), k=2, xtol=atol, rtol=rtol))
+    exp = [
+        tb.astensor(_toms748(f, *best_bracket(i), args=(i), k=2, xtol=atol, rtol=rtol))
+        for i in range(1, 6)
+    ]
+    return obs, exp
+
+
 def upperlimit(data, model, scan, level=0.05, return_results=False):
```

Comment:
hi @beojan couldn't we do `scan = 'auto'` and trigger TOMS here?

i.e if scan is `scan = np.linspace(0,...1)` or any other kind of array honor that ... if `scan='auto'` run TOMS

Reply:
if this works, we'd probably need to return `obs,exp,mus` (where mus=scan in the non-TOMS case) in order to communicate back where the evluation happened

Reply:
Added.